### PR TITLE
fix(core): respect custom proxy_access_log

### LINKED ISSUE #12061: overriding proxy_access_log with different log_format no longer works in Kong 3.5.0
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

Kong 3.5.0

### Current Behavior

We use a custom access log format since years with Kong by adding a custom-nginx.conf and include it via `KONG_NGINX_HTTP_INCLUDE` env var
If we start Kong 3.5.0 with the environment variable `KONG_PROXY_ACCESS_LOG` set to `/dev/stdout custom_proxy_format` Kong fails to start.

```
Error: could not prepare Kong prefix at /usr/local/kong: nginx configuration is invalid (exit code 1):
nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:7
nginx: [emerg] invalid parameter "kong_log_format" in /usr/local/kong/nginx-kong.conf:89
nginx: configuration file /usr/local/kong/nginx.conf test failed


  Run with --v (verbose) or --vv (debug) for more details
```

This happens due to this changed line in Kong release 3.5.0
https://github.com/Kong/kong/commit/6890007457c1c974f953ca442c0eed75458bb6d6#diff-d2fc92c2f1cd1ea7d01c13c68b069cbe9927088d2040d5c096906a93cf95bbc5R88

### Expected Behavior

same behavior as previous Kong versions.

### Steps To Reproduce

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by outsinre**
### Summary

Kong now has a fixed access log format `kong_log_format` that prevents customization. Introduced by #11663.

Resolves #12061.

If the `proxy_access_log` has a log format specified, then replace `kong_log_format` with the custom value.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [n/a] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* replace `kong_log_format` with custom value set by end users.
* update tests to be more robust.

### Issue reference

Fix #_[FTI-5580]_

[FTI-5580]: https://konghq.atlassian.net/browse/FTI-5580?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by outsinre** [comment]
A similar bug to #11712.

File: kong/cmd/utils/prefix_handler.lua [hunk 1]
```diff
@@ -248,6 +249,19 @@ local function compile_conf(kong_config, conf_template, template_env_inject)
...
+  local kong_proxy_access_log = kong_config.proxy_access_log
+  if kong_proxy_access_log ~= "off" then
+    compile_env.proxy_access_log_enabled = true
+  end
+  if kong_proxy_access_log then
+    if (not lfs.attributes(kong_proxy_access_log)) and
+       (not lfs.attributes(kong_config.prefix .. "/" .. kong_proxy_access_log))
+    then
+      log.debug(table.concat({"'proxy_access_log' is a custom value:", kong_proxy_access_log}, " "))
+      compile_env.proxy_access_log_custom = true
```
Comment (samugi) [inline kong/cmd/utils/prefix_handler.lua:261]: If I understand this correctly it is trying to detect whether the `access_log` was configured using some custom format like: `access_log /tmp/access.log custom_format;`

It seems this code will return true when passing `KONG_PROXY_ACCESS_LOG=$(pwd)/some_file.log` if the target file (`some_file.log`) didn't already exist, should we rather detect if a _format_ exists after the path and only then consider it custom?

Comment (outsinre) [reply kong/cmd/utils/prefix_handler.lua:261]: @samugi nice catch.

To determine if a *format* exist or not, we then use *space* to split the `access_log`.

Previously, I worry if the `some_file.log` may contain white spaces, but seems it is the user's responsibility to set a correct filename (with quotes).

Let me update.

Comment (samugi) [reply kong/cmd/utils/prefix_handler.lua:261]: that sounds reasonable to me

Comment (outsinre) [reply kong/cmd/utils/prefix_handler.lua]: @samugi updated.

File: kong/cmd/utils/prefix_handler.lua [hunk 2]
```diff
@@ -248,6 +248,21 @@ local function compile_conf(kong_config, conf_template, template_env_inject)
...
+    compile_env.proxy_access_log_enabled = true
+  end
+  if kong_proxy_access_log then
+    local _, custom_format = string.match(kong_proxy_access_log, "^(%S+)%s(%S+)")
+    if custom_format then
+      compile_env.proxy_access_log_custom = true
+
+      if not string.match(kong_config.nginx_http_log_format or "", "^" .. custom_format) then
+        log.debug(table.concat({"a custom access_log format is set: '", custom_format, "', but not defined"}))
+      end
```
Comment (outsinre) [inline kong/cmd/utils/prefix_handler.lua:262]: If the user sets a custom format but does not define it, kong cannot start, which is expected.

We print a debug log for this case.

File: kong/cmd/utils/prefix_handler.lua [hunk 3]
```diff
@@ -248,6 +248,21 @@ local function compile_conf(kong_config, conf_template, template_env_inject)
...
+  local kong_proxy_access_log = kong_config.proxy_access_log
+  if kong_proxy_access_log ~= "off" then
+    compile_env.proxy_access_log_enabled = true
+  end
+  if kong_proxy_access_log then
+    local _, custom_format = string.match(kong_proxy_access_log, "^(%S+)%s(%S+)")
+    if custom_format then
+      compile_env.proxy_access_log_custom = true
+
+      if not string.match(kong_config.nginx_http_log_format or "", "^" .. custom_format) then
```
Comment (samugi) [inline kong/cmd/utils/prefix_handler.lua:260]: wouldn't this prevent from using any unescaped magic character in the custom format name, like `test-format`?

Comment (outsinre) [reply kong/cmd/utils/prefix_handler.lua]: @samugi make sense.

A format name with a *hyphen* will print the debug message, but does not preventing Kong start. That debug message is a false negative.

Updated with a new commit, so you can verify.

I tested it locally with a real example and it works.

Comment (samugi) [reply kong/cmd/utils/prefix_handler.lua]: @outsinre sorry for a lot of back and forth on this, how about using `string.find` with the `plain` flag set to `true` and look for `"log_format " .. custom_format`?

Comment (outsinre) [reply kong/cmd/utils/prefix_handler.lua]: @samugi awesome idea. So we do not have to manipulate the custom format. Let me try.

File: kong/cmd/utils/prefix_handler.lua [hunk 4]
```diff
@@ -248,6 +248,18 @@ local function compile_conf(kong_config, conf_template, template_env_inject)
     }
   }
 
+  local kong_proxy_access_log = kong_config.proxy_access_log
+  if kong_proxy_access_log ~= "off" then
+    compile_env.proxy_access_log_enabled = true
+  end
+  if kong_proxy_access_log then
+    -- example: 'logs/some-file.log apigw_json'
```
Comment (tzssangglass) [inline kong/cmd/utils/prefix_handler.lua:256]:
```text
-- example: proxy_access_log = logs/some-file.log apigw_json
```

is it more clear?

File: kong/cmd/utils/prefix_handler.lua [hunk 5]
```diff
@@ -248,6 +248,18 @@ local function compile_conf(kong_config, conf_template, template_env_inject)
...
+    local _, custom_format_name = string.match(kong_proxy_access_log, "^(%S+)%s(%S+)")
+    if custom_format_name then
+      compile_env.proxy_access_log_custom = true
```
Comment (tzssangglass) [inline kong/cmd/utils/prefix_handler.lua:259]:
```suggestion
      compile_env.custom_proxy_access_log = true
```

File: spec/01-unit/04-prefix_handler_spec.lua [hunk 6]
```diff
@@ -485,35 +485,71 @@ describe("NGINX conf compiler", function()
     end)
 
     describe("injected NGINX directives", function()
-      it("injects proxy_access_log directive", function()
-        local conf = assert(conf_loader(nil, {
+      it("#test injects proxy_access_log directive", function()
```
Comment (samugi) [inline spec/01-unit/04-prefix_handler_spec.lua:488]:
```suggestion
      it("injects proxy_access_log directive", function()
```

File: spec/01-unit/04-prefix_handler_spec.lua [hunk 7]
```diff
@@ -485,35 +485,71 @@ describe("NGINX conf compiler", function()
...
+        nginx_conf = prefix_handler.compile_kong_stream_conf(conf)
+        assert.matches("access_log%slogs/access.log%sbasic;", nginx_conf)
+
+        conf = assert(conf_loader(nil, {
+          proxy_access_log = "/dev/stdout apigw-json",
+          nginx_http_log_format = 'not-exist "$kong_request_id"',
+          stream_listen = "0.0.0.0:9100",
+          nginx_stream_tcp_nodelay = "on",
+        }))
+        nginx_conf = prefix_handler.compile_kong_conf(conf)
```
Comment (samugi) [inline spec/01-unit/04-prefix_handler_spec.lua:527]: this one's compiling an invalid config that would eventually prevent Kong from starting up, right? It is expected, but maybe a little comment above this chunk would help make it more clear.

**Review by samugi — APPROVED** [review]
LGTM, a couple of minor comments on the tests.

File: kong/cmd/utils/prefix_handler.lua (hunk 5 above)
Comment (samugi) [reply kong/cmd/utils/prefix_handler.lua:259]: ^ just careful to update this everywhere if you commit the change :)

File: spec/01-unit/04-prefix_handler_spec.lua (hunk 7 above)
Comment (outsinre) [reply spec/01-unit/04-prefix_handler_spec.lua:527]: @samugi yep. Previously we print a log or return nil, err.

Let me add a comment.

File: spec/01-unit/04-prefix_handler_spec.lua (hunk 6 above)
Comment (outsinre) [reply spec/01-unit/04-prefix_handler_spec.lua]: Done

File: kong/cmd/utils/prefix_handler.lua (hunk 5 above)
Comment (outsinre) [reply kong/cmd/utils/prefix_handler.lua]: Done

File: kong/cmd/utils/prefix_handler.lua (hunk 4 above)
Comment (outsinre) [reply kong/cmd/utils/prefix_handler.lua]: Done.

File: spec/01-unit/04-prefix_handler_spec.lua (hunk 7 above)
Comment (outsinre) [reply spec/01-unit/04-prefix_handler_spec.lua:527]: Done.

**Review by tzssangglass — APPROVED** [review]
LGTM

**Pull request merged**